src/Client.ts

```diff
@@ -52,29 +51,81 @@ export default class Client extends EventEmitter {
 	 * @param host - hostname or IP address of daemon
 	 * @param port - port number of daemon on the host
 	 */
-	introduce(host: string, port: number): void {
-		const client = connect(`http://${hostAndPort(host, port)}`)
+	public introduce(host: string, port: number): void {
+		const client = connect(`http://${Client.hostAndPort(host, port)}`)
 		client.on('error', (err) => this.emit('error', err))
-		this.available(client)
+		this.setAvailable(client)
 	}
 
-	private available(client: ClientHttp2Session): void {
-		const job = this.queue.pop()
-		if (!job) {
-			client.close(() => this.emit('done'))
-			return
+	/**
+	 * Mark a daemon as available.
+	 *
+	 * @param daemon - The daemon that is available.
+	 */
+	private setAvailable(daemon: ClientHttp2Session): void {
+		this.availableDaemons.add(daemon)
+		this.checkJobsAndAssign()
+	}
+
+	/**
+	 * Tries to assign the next available job to a daemon.
+	 * Closes all daemons if all jobs are finished.
+	 */
+	private checkJobsAndAssign(): void {
+		const job = this.jobOrderer.popNextJob()
+
+		if (job && this.availableDaemons.size > 0) {
+			const daemon = Array.from(this.availableDaemons.values())[0]
+			if (daemon) {
+				this.assignJobToDaemon(job, daemon)
+			} else {
+				throw 'skjdfhsadkfl'
+			}
+		} else if (this.jobOrderer.isDone()) {
+			this.closeAllDaemonsAndFinish()
 		}
+	}
 
-		const req = client.request({ ':path': `/${job}` })
-		client.on('error', (err) => this.emit('error', err))
+	/**
+	 * Closes all daemons.
+	 * Emits 'done' once all daemons are closed.
+	 */
+	private closeAllDaemonsAndFinish() {
+		this.availableDaemons.forEach((daemon) =>
+			daemon.close(() => {
+				this.availableDaemons.delete(daemon)
+				if (this.availableDaemons.size == 0) {
+					this.emit('done')
+				}
+			}),
+		)
+	}
+
+	/**
+	 * Asks the daemon to work on the job.
+	 * Sends HTTP request and handles response.
+	 *
+	 * @param job - The job to assign.
+	 * @param daemon - The daemon to which to assign the job.
+	 */
+	private assignJobToDaemon(job: Job, daemon: ClientHttp2Session) {
+		const request = daemon.request({ ':path': `/${job.name}` })
+		daemon.on('error', (err) => this.emit('error', err))
 
 		let data = ''
-		req.setEncoding('utf8')
-		req.on('data', (chunk) => (data += chunk))
+		request.setEncoding('utf8')
+		request.on('data', (chunk) => (data += chunk))
+
+		request.on('end', () => {
+			if (data == 'failed') {
+				this.jobOrderer.reportFailedJob(job)
+			} else {
+				this.jobOrderer.reportCompletedJob(job)
+			}
 
-		req.on('end', () => {
-			this.emit('progress', job, data)
-			this.available(client)
+			this.emit('progress', job.toString(), data)
+			this.setAvailable(daemon)
+			this.checkJobsAndAssign()
```

Comment:
This gets called twice (once within setAvailable). It's probably best to make setAvailable not call this, or change its name to make clear that it does.